_posts/2025-04-25-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,399 @@
+---
+layout: post
+title: 'Getting ready for secure MCP with Quarkus MCP Server'
+date: 2025-04-25
+tags: ai mcp security
+synopsis: 'Explain how MCP clients can access secure Quarkus MCP SSE servers with access tokens'
+author: sberyozkin
+---
+:imagesdir: /assets/images/posts/secure_mcp_sse_server
+
+== Introduction
+
+https://modelcontextprotocol.io/specification/2025-03-26[The latest version of the Model Context Protocol (MCP) specification] introduces an https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization[authorization] flow.
+
+While it will take a bit of time for the new MCP specification to be widely supported, you can already add authentication to client and server following the https://modelcontextprotocol.io/specification/2024-11-05[previous MCP version].
+
+Indeed, you can use any MCP client that can receive an access token and pass it to the MCP server.
+
+In this post, we will create a https://github.com/quarkiverse/quarkus-mcp-server[Quarkus MCP SSE Server] that requires authentication to access its tools.
+
+We will use Keycloak to login and use a Keycloak JWT access token to access the server with `Quarkus MCP SSE Server Dev UI` in the dev mode.
+
+We will use GitHub to login and use a GitHub binary access token to access the server in the prod mode with both https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the `curl` tools.
```

Comment:
```suggestion
Secondly we will show how to use GitHub OpenID to login and use a GitHub binary access token to access the server in the prod mode with both https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the `curl` tools.
```

Reply:
not sure if OpenID is the correct term but idea is to just highlight its two different ways, keycloak not the only option and openid/oauth is that common enabler.

Reply:
Sure, @maxandersen, I'll tune it a tiny bit and will apply

Reply:
OpenID is fine but is more general SSO term, OpenId Connect, being built on top of OAuth2, is an OpenID in action

Reply:
GitHub and many other social providers are pure OAuth2, they don't offer ID tokens